kmall.py

```diff
@@ -3425,37 +3481,201 @@ def seek_next_startbyte(self, file_length, start_ptr=0):
+        """
+        Read the file and return a dict of the wanted records/fields according to recs_categories.  If start_ptr/end_ptr
+        is provided, start and end at those byte offsets.
+        """
+        wanted_records = list(recs_categories.keys())
+        recs_to_read = {}
+
         if self.FID is None:
             self.OpenFiletoRead()
 
         filelen = self._initialize_sequential_read(start_ptr, end_ptr)
         if start_ptr:
             self.seek_next_startbyte(filelen, start_ptr=start_ptr)
 
-        if self.FID.tell() >= start_ptr + filelen:
-            self.eof = True
-        else:
-            pass
+        while not self.eof:
+            if self.FID.tell() >= start_ptr + filelen:
+                self.eof = True
+            self.decode_datagram()
+            if self.datagram_ident not in wanted_records:
+                self.skip_datagram()
+                continue
+            self.read_datagram()
+            rec = self.datagram_data
+            recs = self._divide_rec(rec)  # split up the MRZ record for multiple sectors, otherwise just returns [rec]
+            for rec in recs:
+                for subrec in recs_categories[self.datagram_ident]:
+                    #  override for nested recs, designated with periods in the recs_to_read dict
+                    if subrec.find('.') > 0:
+                        if len(subrec.split('.')) == 3:
+                            rec_key = subrec.split('.')[2]
+                            tmprec = rec[subrec.split('.')[0]][subrec.split('.')[1]][rec_key]
+                        else:
+                            rec_key = subrec.split('.')[1]
+                            tmprec = rec[subrec.split('.')[0]][rec_key]
+                    else:
+                        rec_key = subrec
+                        tmprec = rec[rec_key]
+
+                    if subrec == 'install_txt':  # str, casting to list splits the string, dont want that
+                        val = [tmprec]
+                    else:
+                        try:  # flow for array/list attribute
+                            val = [np.array(tmprec)]
+                        except TypeError:  # flow for float/int attribute
+                            val = [tmprec]
+
+                    # generate new list or append to list for each rec of that dgram type found
+                    for translated in recs_categories_translator[self.datagram_ident][subrec]:
+                        if translated[0] not in recs_to_read:
+                            recs_to_read[translated[0]] = {}
+                        if translated[1] not in recs_to_read[translated[0]]:
+                            recs_to_read[translated[0]][translated[1]] = val
+                        else:
+                            recs_to_read[translated[0]][translated[1]].extend(val)
+        recs_to_read = self._finalize_records(recs_to_read)
+        return recs_to_read
+
+    def translate_yawpitch(self, arr):
```

Comment:
In this function, and the others that do this kind of translation, I think it would be good to name them something like translate_yawpitchstab_tonumeric(), so it is a bit more descriptive. Also in the doc strings for these, it would be good to try to briefly capture why one would want to do the translation.

Reply:
done.